File: library/src/main/java/io/ashdavies/rx/rxfirebase/ChildEvent.java
package io.ashdavies.rx.rxfirebase;

import android.support.annotation.Nullable;

import com.google.firebase.database.DataSnapshot;

@SuppressWarnings("WeakerAccess")
public final class ChildEvent {
  private final DataSnapshot snapshot;
  private final Type type;

  @Nullable
  private final String previous;

  private ChildEvent(DataSnapshot snapshot, Type type, @Nullable String previous) {
    this.snapshot = snapshot;
    this.type = type;

    this.previous = previous;
  }

  public static ChildEvent create(DataSnapshot dataSnapshot, Type type) {
    return new ChildEvent(dataSnapshot, type, null);
  }

  public static ChildEvent create(DataSnapshot dataSnapshot, Type type, String previous) {
    return new ChildEvent(dataSnapshot, type, previous);
  }

  public final DataSnapshot snapshot() {
    return snapshot;
  }

  public final Type type() {
    return type;
  }

  public final String previous() {
    return previous;
  }

  public enum Type {
    CHILD_ADDED,
    CHILD_CHANGED,
    CHILD_REMOVED,
    CHILD_MOVED
  }
}

File: src/main/kotlin/io/ashdavies/rx/rxfirebase/ChildEvent.kt
package io.ashdavies.rx.rxfirebase

import com.google.firebase.database.DataSnapshot

data class ChildEvent(val snapshot: DataSnapshot, val type: Type, val previous: String?) {

  constructor(snapshot: DataSnapshot, type: Type) : this(snapshot, type, null)

  enum class Type {
    CHILD_ADDED,
    CHILD_CHANGED,
    CHILD_REMOVED,
    CHILD_MOVED
  }
}
